### PR TITLE
Various improvements to http_server_fast.cpp:

### PR DESCRIPTION
**Pull request by chriskohlhoff**
- Receive request in a single read (addresses #574)
- Use fields_alloc for response
- Fix command line usage information
- Add command line option to spin the io_service

### CODE REVIEW

File: example/http-server-fast/http_server_fast.cpp [hunk 1]
```diff
@@ -58,24 +56,26 @@ class http_worker
     std::string doc_root_;
 
     // The socket for the currently connected client.
-    tcp::socket socket_;
+    tcp::socket socket_{acceptor_.get_io_service()};
```
Comment (vinniefalco) [inline example/http-server-fast/http_server_fast.cpp]: is this C++11?

File: example/http-server-fast/http_server_fast.cpp [hunk 2]
```diff
@@ -58,24 +56,26 @@ class http_worker
...
     // The parser for reading the requests
-    using alloc_type = fields_alloc<char>;
-    alloc_type alloc_;
-    boost::optional<http::request_parser<request_body_t, alloc_type>> parser_;
+    boost::optional<http::request_parser<request_body_t, alloc_t>> parser_;
 
     // The timer putting a time limit on requests.
-    boost::asio::basic_waitable_timer<std::chrono::steady_clock> request_deadline_;
+    boost::asio::basic_waitable_timer<std::chrono::steady_clock> request_deadline_{
+        acceptor_.get_io_service(), std::chrono::steady_clock::time_point::max()};
```
Comment (vinniefalco) [inline example/http-server-fast/http_server_fast.cpp]: I think we need to parenthesis
```
(std::chrono::steady_clock::time_point::max)()
```

**Comment by vinniefalco** [comment]
I'm a little confused about the handling of the timer. The docs say that `expires_from_now` cancels pending waits. But in http_server_fast.cpp we call `expires_from_now` followed by `read_header` and never establish a wait. Is this correct?

File: example/http-server-fast/http_server_fast.cpp (hunk 1 above)
Comment (chriskohlhoff) [inline example/http-server-fast/http_server_fast.cpp]: yep

File: example/http-server-fast/http_server_fast.cpp (hunk 2 above)
Comment (chriskohlhoff) [inline example/http-server-fast/http_server_fast.cpp]: aaaaah Windows

**Comment by chriskohlhoff** [comment]
> On 4 Jul 2017, at 12:39 am, Vinnie Falco <notifications@github.com> wrote:
> 
> I'm a little confused about the handling of the timer. The docs say that expires_from_now cancels pending waits. But in http_server_fast.cpp we call expires_from_now followed by read_header and never establish a wait. Is this correct?

The check_deadline function "loop" is responsible for doing the waiting. It will wake up and then go to sleep again on the new timeout.

**Comment by vinniefalco** [comment]
@chriskohlhoff Very nice way to abuse.. I mean, use :) the timer! I will be stealing that.

**Comment by chriskohlhoff** [comment]
On Tue, Jul 4, 2017, at 08:29 AM, Vinnie Falco wrote:
> @chriskohlhoff Very nice way to abuse.. I mean, use :) the timer! I will
> be stealing that.

The Boost.Asio example program example/cpp03/timeouts/server.cpp
illustrates this approach in more detail.

**Comment by vinniefalco** [comment]
Always nice to find out I'm behind the curve on a 7 year old blog post

**Pull request closed without merging**